src/pkg/packager/publish.go

```diff
@@ -0,0 +1,407 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2021-Present The Zarf Authors
+
+// Package packager contains functions for interacting with, managing and deploying Zarf packages.
+package packager
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/defenseunicorns/zarf/src/config"
+	"github.com/defenseunicorns/zarf/src/pkg/message"
+	"github.com/defenseunicorns/zarf/src/pkg/utils"
+	"github.com/mholt/archiver/v3"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/content/file"
+	"oras.land/oras-go/v2/registry"
+)
+
+// ZarfLayerMediaType<Extension> is the media type for Zarf layers.
+const (
+	ZarfLayerMediaTypeTarZstd = "application/vnd.zarf.layer.v1.tar+zstd"
+	ZarfLayerMediaTypeTarGzip = "application/vnd.zarf.layer.v1.tar+gzip"
+	ZarfLayerMediaTypeYaml    = "application/vnd.zarf.layer.v1.yaml"
+	ZarfLayerMediaTypeJSON    = "application/vnd.zarf.layer.v1.json"
+	ZarfLayerMediaTypeTxt     = "application/vnd.zarf.layer.v1.txt"
+	ZarfLayerMediaTypeUnknown = "application/vnd.zarf.layer.v1.unknown"
+)
+
+// parseZarfLayerMediaType returns the Zarf layer media type for the given filename.
+func parseZarfLayerMediaType(filename string) string {
+	// since we are controlling the filenames, we can just use the extension
+	switch filepath.Ext(filename) {
+	case ".zst":
+		return ZarfLayerMediaTypeTarZstd
+	case ".gz":
+		return ZarfLayerMediaTypeTarGzip
+	case ".yaml":
+		return ZarfLayerMediaTypeYaml
+	case ".json":
+		return ZarfLayerMediaTypeJSON
+	case ".txt":
+		return ZarfLayerMediaTypeTxt
+	default:
+		return ZarfLayerMediaTypeUnknown
+	}
+}
+
+// Publish publishes the package to a registry
+//
+// This is a wrapper around the oras library
+// and much of the code was adapted from the oras CLI - https://github.com/oras-project/oras/blob/main/cmd/oras/push.go
+//
+// Authentication is handled via the Docker config file created w/ `zarf tools registry login`
+func (p *Packager) Publish() error {
+	p.cfg.DeployOpts.PackagePath = p.cfg.PublishOpts.PackagePath
+	if err := p.loadZarfPkg(); err != nil {
+		return fmt.Errorf("unable to load the package: %w", err)
+	}
+
+	/*
+		Structure:
+		.
+		├── components/
+		│   ├── component1.tar
+		│   ├── component2.tar
+		│   └── [...].tar
+		├── images/
+		│   ├── index.json
+		│   ├── oci-layout
+		│   ├── blobs/
+		│   │   ├── sha256/
+		│   │   │   └── [...]
+		├── checksums.txt (optional)
+		├── sboms.tar.zst
+		└── zarf.yaml
+	*/
```

Comment:
This would probably be better if it linked to docs somewhere instead of code comments